### PR TITLE
Connector conditional put

### LINKED ISSUE #2688: [Question] KafkaConnector getting updated even when resource not modified
We are experiencing an issue when our KafkaConnectors are getting updated every 120 seconds.

Somehow stored connector status is never match desired status.
```
2020-03-12 03:22:47 DEBUG StatusDiff:43 - Current Status path /connectorStatus/connector/state has value "UNASSIGNED"
2020-03-12 03:22:47 DEBUG StatusDiff:44 - Desired Status path /connectorStatus/connector/state has value "RUNNING"
``` 
Always followed by this on next reconciliation timer tick 
```
2020-03-12 03:24:39 DEBUG StatusDiff:43 - Current Status path /connectorStatus/connector/state has value "RUNNING"
2020-03-12 03:24:39 DEBUG StatusDiff:44 - Desired Status path /connectorStatus/connector/state has value "UNASSIGNED"
```

Not sure how to break this loop and whether it is a bug or our misconfiguration.

Environment:
- Strimzi version: 0.16.2
- Kubernetes cluster: 1.14
- Infrastructure: EKS

Do let me know if any additional data required.

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Bugfix

### Description

This PR fixes #2688 and adds a test for it. The fix is pretty simple: Get the connector config from the REST API and only update if the config differs. To do this I had to add a new method to the REST API client.

The PR also adds an integration test which runs the `reconcileConnector()` method against a real Kafka Connect cluster. 

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 1]
```diff
@@ -333,40 +338,95 @@ protected KafkaConnectApi getKafkaConnectApi() {
...
+     * @param connectorName The connector name.
+     * @param connectorSpec The desired connector spec.
+     * @return A Future whose result, when successfully completed, is a map of the current connector state.
+     */
+    protected Future<Map<String, Object>> maybeCreateOrUpdateConnector(Reconciliation reconciliation, String host, KafkaConnectApi apiClient,
+                                                                       String connectorName, KafkaConnectorSpec connectorSpec) {
+        return apiClient.getConnectorConfig(new BackOff(200L, 2, 6), host, port, connectorName).compose(
+            config -> {
+                if (!needsReconfiguring(reconciliation, connectorName, connectorSpec, config)) {
+                    log.info("{}: Connector {} exists and has desired config, {}=={}", reconciliation, connectorName, connectorSpec.getConfig(), config);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java]: Should this be really `info` level?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 2]
```diff
@@ -333,40 +338,95 @@ protected KafkaConnectApi getKafkaConnectApi() {
...
-                        return apiClient.resume(host, KafkaConnectCluster.REST_API_PORT,
-                                connectorName)
-                                .compose(ignored ->
-                                        apiClient.status(host, KafkaConnectCluster.REST_API_PORT,
-                                                connectorName));
-
-                    } else {
-                        return Future.succeededFuture(status);
-                    }
+                    log.info("{}: Connector {} exists but does not have desired config, {}!={}", reconciliation, connectorName, connectorSpec.getConfig(), config);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java]: Same as above.

**Comment by tombentley** [comment]
@strimzi-ci run tests testcase=ConnectST

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java [hunk 3]
```diff
@@ -333,40 +338,95 @@ protected KafkaConnectApi getKafkaConnectApi() {
...
 
+    private boolean needsReconfiguring(Reconciliation reconciliation, String connectorName, KafkaConnectorSpec connectorSpec, Map<String, Object> actual) {
+        // The actual which comes from Connect API includes tasks.max, connector.class and name,
+        // which connectorSpec.getConfig() does not
+        Map<String, Object> desired = log.isDebugEnabled() ? new TreeMap<>(connectorSpec.getConfig()) : new HashMap<>(connectorSpec.getConfig());
+        desired.put("tasks.max", connectorSpec.getTasksMax().toString());
+        desired.put("name", connectorName);
+        desired.put("connector.class", connectorSpec.getClassName());
+        if (log.isDebugEnabled()) {
+            log.info("{}: Desired: {}", reconciliation, desired);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java]: Given the `isDebugEnabled` I guess this should definitely be `debug`?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java [hunk 4]
```diff
@@ -89,13 +188,16 @@ public KafkaConnectApiImpl(Vertx vertx) {
                         response.bodyHandler(buffer -> {
                             ObjectMapper mapper = new ObjectMapper();
                             try {
-                                result.complete(mapper.readValue(buffer.getBytes(), Map.class));
+                                Map t = mapper.readValue(buffer.getBytes(), Map.class);
+                                log.info("Got {} response to PUT request to {}: {}", response.statusCode(), path, t);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java]: `debug` ot `trace` level I guess?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java [hunk 5]
```diff
@@ -89,13 +188,16 @@ public KafkaConnectApiImpl(Vertx vertx) {
...
+                                result.complete(t);
                             } catch (IOException e) {
                                 result.fail(new ConnectRestException(response, "Could not deserialize response: " + e));
                             }
                         });
                     } else {
                         // TODO Handle 409 (Conflict) indicating a rebalance in progress
+                        log.info("Got {} response to PUT request to {}", response.statusCode(), path);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java]: Again ...

**Comment by tombentley** [comment]
@strimzi-ci run tests testcase=ConnectST

**Comment by tombentley** [comment]
@strimzi-ci run tests testcase=ConnectST

**Pull request merged**